.dotnet/src/Custom/Chat/StreamingChatUpdate.Serialization.cs

```diff
@@ -0,0 +1,181 @@
+
+    BinaryData IPersistableModel<IEnumerable<StreamingChatUpdate>>.Write(ModelReaderWriterOptions options)
+        => ModelSerializationHelpers.SerializeInstance<IEnumerable<StreamingChatUpdate>, StreamingChatUpdate>(this, options);
+
+    string IPersistableModel<IEnumerable<StreamingChatUpdate>>.GetFormatFromOptions(ModelReaderWriterOptions options) => "J";
+
+    internal static IEnumerable<StreamingChatUpdate> DeserializeStreamingChatUpdates(
+        JsonElement sseDataJson,
+        ModelReaderWriterOptions options = default)
+    {
+        List<StreamingChatUpdate> results = [];
+        if (sseDataJson.ValueKind == JsonValueKind.Null)
+        {
+            return results;
+        }
+        string id = default;
+        DateTimeOffset created = default;
+        string systemFingerprint = null;
+        foreach (JsonProperty property in sseDataJson.EnumerateObject())
+        {
+            if (property.NameEquals("id"u8))
+            {
+                id = property.Value.GetString();
+                continue;
+            }
+            if (property.NameEquals("created"u8))
+            {
+                created = DateTimeOffset.FromUnixTimeSeconds(property.Value.GetInt64());
+                continue;
+            }
+            if (property.NameEquals("system_fingerprint"))
+            {
+                systemFingerprint = property.Value.GetString();
+                continue;
+            }
+            if (property.NameEquals("choices"u8))
+            {
+                foreach (JsonElement choiceElement in property.Value.EnumerateArray())
+                {
+                    ChatRole? role = null;
+                    string contentUpdate = null;
+                    string functionName = null;
+                    string functionArgumentsUpdate = null;
+                    int choiceIndex = 0;
+                    ChatFinishReason? finishReason = null;
+                    List<StreamingToolCallUpdate> toolCallUpdates = [];
+                    ChatLogProbabilityCollection logProbabilities = new([]);
+
+                    foreach (JsonProperty choiceProperty in choiceElement.EnumerateObject())
+                    {
+                        if (choiceProperty.NameEquals("index"u8))
+                        {
+                            choiceIndex = choiceProperty.Value.GetInt32();
+                            continue;
+                        }
+                        if (choiceProperty.NameEquals("finish_reason"u8))
+                        {
+                            if (choiceProperty.Value.ValueKind == JsonValueKind.Null)
+                            {
+                                finishReason = null;
+                                continue;
+                            }
+                            finishReason = choiceProperty.Value.GetString() switch
```

Comment:
this is lower pri, but would it be better to stay in UTF8 span world and use a sequence of if statements (like for top level properties)